File: apps/dashboard/lib/agents.ts
/**
 * Agent workflow API client — calls the public /agents/demo-workflow endpoint.
 *
 * The backend runs each agent (pure computation, no DB/auth required) and
 * returns the full structured output of every agent for rendering proof.
 */

import api from "./api";

/** Raw structured output of a single agent run. */
export interface AgentOutput {
  name: string;
  display: string;
  desc: string;
  success: boolean;
  processing_time_ms: number;
  /** Parsed JSON from the API — rendered by AgentOutputRenderer. */
  output: Record<string, any>;
  suggestions: string[];
}

/** Response shape of POST /agents/demo-workflow. */
export interface DemoWorkflowResponse {
  success: boolean;
  workflow_id: string;
  keyword: string;
  niche: string;
  agents: AgentOutput[];
  master: {
    seo_score: number;
    quality_score: number;
    total_processing_time_ms: number;
    agents_run: number;
  };
}

export interface WorkflowOptions {
  keyword: string;
  niche?: string;
  audience?: string;
  tone?: string;
  goal?: string;
  /** Subset of agents to run, e.g. ["seo", "content"]. */
  agents?: string[];
}

/** Canonical agent metadata for the dashboard UI. */
export const AGENT_META: Record<
  string,
  { display: string; icon: string; gradient: string; ring: string }
> = {
  trend: {
    display: "Trend Agent",
    icon: "trending-up",
    gradient: "from-blue-500/15 to-cyan-500/5",
    ring: "ring-blue-500/30",
  },
  seo: {
    display: "SEO Agent",
    icon: "search",
    gradient: "from-emerald-500/15 to-teal-500/5",
    ring: "ring-emerald-500/30",
  },
  content: {
    display: "Content Agent",
    icon: "file-text",
    gradient: "from-amber-500/15 to-orange-500/5",
    ring: "ring-amber-500/30",
  },
  design: {
    display: "Design Agent",
    icon: "palette",
    gradient: "from-pink-500/15 to-rose-500/5",
    ring: "ring-pink-500/30",
  },
  quality: {
    display: "Quality Agent",
    icon: "check-circle",
    gradient: "from-red-500/15 to-rose-500/5",
    ring: "ring-red-500/30",
  },
  scheduler: {
    display: "Scheduler Agent",
    icon: "clock",
    gradient: "from-indigo-500/15 to-blue-500/5",
    ring: "ring-indigo-500/30",
  },
  analytics: {
    display: "Analytics Agent",
    icon: "bar-chart-3",
    gradient: "from-teal-500/15 to-emerald-500/5",
    ring: "ring-teal-500/30",
  },
  strategy: {
    display: "Strategy Agent",
    icon: "lightbulb",
    gradient: "from-violet-500/15 to-purple-500/5",
    ring: "ring-violet-500/30",
  },
  master: {
    display: "Master Agent",
    icon: "cpu",
    gradient: "from-purple-500/15 to-fuchsia-500/5",
    ring: "ring-purple-500/30",
  },
};

export const AGENT_ORDER = [
  "trend",
  "seo",
  "content",
  "design",
  "quality",
  "scheduler",
  "analytics",
  "strategy",
];

/** Run the demo workflow against the backend. Throws ApiError on failure. */
export async function runDemoWorkflow(
  options: WorkflowOptions,
): Promise<DemoWorkflowResponse> {
  return api.post<DemoWorkflowResponse>("/agents/demo-workflow", {
    keyword: options.keyword,
    niche: options.niche || "general",
    audience: options.audience || "",
    tone: options.tone || "professional",
    goal: options.goal || "engagement",
    agents: options.agents || AGENT_ORDER,
  });
}

File: apps/api/app/api/agents.py
"""
AI Multi-Agent System API endpoints.

Exposes the Master Agent and individual agents for orchestration.
"""

from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.api.auth import get_current_user
from app.models.user import User
from app.agents.base import AgentContext
from app.agents.master.agent import MasterAgent

router = APIRouter()
master_agent = MasterAgent()


class DemoWorkflowRequest(BaseModel):
    """Request for the public demo workflow — runs any subset of the 9 agents."""
    keyword: str
    niche: str = "general"
    audience: str = ""
    tone: str = "professional"
    goal: str = "engagement"
    agents: list[str] = [
        "trend", "seo", "content", "design", "quality",
        "scheduler", "analytics", "strategy",
    ]


ALL_AGENTS = {
    "trend": {"display": "Trend Agent", "desc": "Trend discovery & opportunities"},
    "seo": {"display": "SEO Agent", "desc": "Keyword research & SEO scoring"},
    "content": {"display": "Content Agent", "desc": "Titles, descriptions, hashtags, CTAs"},
    "design": {"display": "Design Agent", "desc": "Graphic specs & design variations"},
    "quality": {"display": "Quality Agent", "desc": "Quality checks & policy review"},
    "scheduler": {"display": "Scheduler Agent", "desc": "Posting windows & queue"},
    "analytics": {"display": "Analytics Agent", "desc": "Performance metrics & reports"},
    "strategy": {"display": "Strategy Agent", "desc": "Growth recommendations & roadmap"},
}


@router.post("/demo-workflow")
async def run_demo_workflow(body: DemoWorkflowRequest):
    """Public endpoint — runs all (or selected) agents without auth or database.

    Used by the dashboard to demonstrate every agent generating real output.
    Each agent returns its full structured output for proof.
    """
    keyword = body.keyword.strip()
    if not keyword:
        raise HTTPException(status_code=400, detail="Keyword is required")

    ctx = AgentContext(
        workflow_id=str(uuid4()),
        keyword=keyword,
        niche=body.niche or "general",
        audience=body.audience,
        tone=body.tone,
        goal=body.goal,
    )

    # Filter to valid agent names, preserving canonical order
    requested = [a for a in ALL_AGENTS if a in body.agents]
    results = []

    for name in requested:
        agent = master_agent.agents[name]
        res = await agent.run(ctx)
        results.append({
            "name": name,
            "display": ALL_AGENTS[name]["display"],
            "desc": ALL_AGENTS[name]["desc"],
            "success": res.success,
            "processing_time_ms": res.processing_time_ms,
            "output": res.output,
            "suggestions": res.suggestions,
        })

        # Propagate context like the Master Agent does
        if name == "trend" and res.success:
            ctx.trend_data = res.output
        elif name == "seo" and res.success:
            ctx.generated_keywords = res.output.get("keywords", [])
            ctx.seo_score = res.output.get("seo_score", 0.0)
        elif name == "content" and res.success:
            ctx.generated_titles = res.output.get("titles", [])
            ctx.generated_descriptions = res.output.get("descriptions", [])
            ctx.generated_hashtags = res.output.get("hashtags", [])
        elif name == "design" and res.success:
            ctx.generated_images = res.output.get("images", [])
        elif name == "quality" and res.success:
            ctx.quality_score = res.output.get("quality_score", 0.0)

    return {
        "success": True,
        "workflow_id": ctx.workflow_id,
        "keyword": keyword,
        "niche": ctx.niche,
        "agents": results,
        "master": {
            "seo_score": ctx.seo_score,
            "quality_score": ctx.quality_score,
            "total_processing_time_ms": round(sum(r["processing_time_ms"] for r in results), 2),
            "agents_run": len(results),
        },
    }

@router.post("/run-workflow")
async def run_full_workflow(
    keyword: str,
    niche: str = "general",
    audience: str = "",
    tone: str = "professional",
    goal: str = "engagement",
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
):
    """Run the complete multi-agent workflow from trend discovery to strategy."""
    context = AgentContext(
        user_id=str(user.id),
        keyword=keyword,
        niche=niche,
        audience=audience,
        tone=tone,
        goal=goal,
    )
    result = await master_agent.execute(context)
    return {
        "success": result.success,
        "workflow_id": result.workflow_id,
        "output": result.output,
        "processing_time_ms": result.processing_time_ms,
        "error": result.error,
    }


@router.post("/trend-scan")
async def run_trend_scan(
    keyword: str,
    niche: str = "general",
    user: User = Depends(get_current_user),
):
    """Run only the trend discovery agent."""
    result = await master_agent.run_trend_scan(keyword, niche, str(user.id))
    return {
        "success": result.success,
        "output": result.output,
        "suggestions": result.suggestions,
        "processing_time_ms": result.processing_time_ms,
    }


@router.post("/generate-content")
async def run_content_generation(
    keyword: str,
    niche: str = "general",
    tone: str = "professional",
    audience: str = "",
    count: int = 5,
    user: User = Depends(get_current_user),
):
    """Run the SEO + Content generation workflow."""
    result = await master_agent.run_content_generation(
        keyword, niche, tone, audience, count, str(user.id)
    )
    return {
        "success": result.success,
        "output": result.output,
        "processing_time_ms": result.processing_time_ms,
    }


@router.get("/agents")
async def list_agents():
    """List all available agents and their capabilities."""
    return {
        "agents": [
            {"name": "Master", "description": "Coordinates all agents and manages workflows"},
            {"name": "Trend", "description": "Discovers trending topics and content opportunities"},
            {"name": "SEO", "description": "Keyword research, SEO optimization, metadata generation"},
            {"name": "Content", "description": "Generates titles, descriptions, hashtags, and CTAs"},
            {"name": "Design", "description": "Creates Pinterest-optimized images and design variations"},
            {"name": "Quality", "description": "Reviews content quality, grammar, and policy compliance"},
            {"name": "Scheduler", "description": "Optimizes publishing schedules and queue management"},
            {"name": "Analytics", "description": "Analyzes performance data and generates reports"},
            {"name": "Strategy", "description": "Generates strategic recommendations and growth roadmap"},
        ],
        "total_agents": 9,
    }